### PR TITLE
[GH-280] Display Energy bar in Battle

### LINKED ISSUE #280: Display Energy bar in battle
Display energy bars for each unit in the battle in a way that is similar to the health bar

Depends on https://github.com/lambdaclass/champions_of_mirra/issues/396

Additional: send a StatOverride message from the backend to the frontend, in order to reset the energy to 0 after using the ultimate skill of a character.

### PR DESCRIPTION
**Pull request by ncontinanza**
Closes #280 

⚠️ Test together with this backend PR: https://github.com/lambdaclass/mirra_backend/pull/564

## Motivation

Until now, we couldn't see the energy accumulated by each unit in the battle. 

## Summary of changes

- Add an energy bar to the unit prefab, right below the health bar.
- Increase the energy level each time the unit attacks with the basic skill, and decrease the bar to zero when the ultimate skill is casted.
- Receive a new message from the backend, StatOverride, to support the case in which we need to set a specific value to a stat, such as setting it to zero when the energy is consumed to perform the ultimate attack. This message will be reused in the near future for other stats.

## How has this been tested?

Start the backend (use [this](https://github.com/lambdaclass/mirra_backend/pull/564) branch!) and fight a battle. You should see a blue energy bar above each unit's head, that increases with each use of the basic skill and decreases with each use of the ultimate skill (that is, after is reaches its top).

## Test additions / changes

List tests added/updated.

## Checklist
- [X] I have tested the changes locally.
- [X] I self-reviewed the changes on GitHub, line by line.
- [ ] Tests have been added/updated.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] I have tested the changes in another devices.
  - [ ] Tested in iOS.
  - [ ] Tested in Android.

### CODE REVIEW

**Review by tkz00 — CHANGES_REQUESTED** [review]
https://github.com/lambdaclass/champions_of_mirra/assets/67983386/6aa45e44-7f73-4a34-8278-8db9edb4b120

I see that after casting their first ultimate of the battle, units stop charging it's energy bar, except sometimes that they charge a little and then it goes away without it being complete.

**Comment by ncontinanza** [comment]
> Screen.Recording.2024-05-03.at.2.54.56.PM.mov 
> I see that after casting their first ultimate of the battle, units stop charging it's energy bar, except sometimes that they charge a little and then it goes away without it being complete.

Thanks! I wasn't limiting the amount to add/substract from the unit's energy, so in some cases I was allowing negative amounts of energy.

**Review by tkz00 — APPROVED** [review]

**Review by tkz00 — DISMISSED** [review]

File: client/Assets/Scripts/Battle/BattleManager.cs [hunk 1]
```diff
@@ -114,6 +119,13 @@ private IEnumerator PlayOutSteps(RepeatedField<Protobuf.Messages.Step> steps)
 					case Protobuf.Messages.Action.ActionTypeOneofCase.Death:
 						battleUnitsUI.Where(battleUnit => battleUnit.SelectedUnit != null).Single(unit => unit.SelectedUnit.id == action.Death.UnitId).DeathFeedback();
 						break;
+					case Protobuf.Messages.Action.ActionTypeOneofCase.EnergyRegen:
+						BattleUnit unit = battleUnitsUI.Single(unit => unit.SelectedUnit.id == action.EnergyRegen.TargetId);
```
Comment (tkz00) [inline client/Assets/Scripts/Battle/BattleManager.cs:123]:
```suggestion
						BattleUnit unit = battleUnitsUI.Single(unit => battleUnit.SelectedUnit != null && unit.SelectedUnit.id == action.EnergyRegen.TargetId);
```

This change allows the battle to play out with teams with less than 6 untis, we should really change this "search" behavior and encapsulate it in a single method.

**Review by tkz00 — CHANGES_REQUESTED** [review]
Found an error when battling with less than 6 units in either team.

File: client/Assets/Scripts/Battle/BattleManager.cs [hunk 2]
```diff
@@ -276,4 +288,20 @@ private List<UIReward> CreateRewardsList()
 		}
 		return rewards;
 	}
+
+	private void StatOverride(Protobuf.Messages.Action action)
+	{
+		BattleUnit target = battleUnitsUI.Single(unit => unit.SelectedUnit.id == action.StatOverride.TargetId);
```
Comment (tkz00) [inline client/Assets/Scripts/Battle/BattleManager.cs:294]:
```suggestion
						BattleUnit target = battleUnitsUI.Single(unit => unit.SelectedUnit != null && unit.SelectedUnit.id == action.StatOverride.TargetId);
```

Sorry, the same problem arrises here, it's the last one, I've checked.

Comment (ncontinanza) [reply client/Assets/Scripts/Battle/BattleManager.cs]: Nice catch!

**Review by tkz00 — APPROVED** [review]

File: client/Assets/Scripts/Battle/BattleManager.cs [hunk 3]
```diff
@@ -276,4 +288,20 @@ private List<UIReward> CreateRewardsList()
...
+	private void StatOverride(Protobuf.Messages.Action action)
+	{
+		BattleUnit target = battleUnitsUI.Single(unit => unit.SelectedUnit != null && unit.SelectedUnit.id == action.StatOverride.TargetId);
+		switch (action.StatOverride.StatAffected.Stat)
+		{
+			case Protobuf.Messages.Stat.Health:
+				target.CurrentHealth = (int)action.StatOverride.StatAffected.Amount;
+				break;
+			case Protobuf.Messages.Stat.Energy:
+				target.CurrentEnergy = Math.Max(0, (int)action.StatOverride.StatAffected.Amount);
```
Comment (lotuuu) [inline client/Assets/Scripts/Battle/BattleManager.cs:301]: Why are we introducing game logic here? If our battle design does not allow us to have negative energy, then it's not a good idea to hide it here, as the battle will be buggy and we will have less ways of finding out!

**Review by lotuuu — CHANGES_REQUESTED** [review]

**Review by lotuuu — APPROVED** [review]

**Review by tkz00 — APPROVED** [review]

**Pull request merged**